mbed_targets/_internal/target_attribute_hierarchy_parsers/accumulating_attribute_parser.py

```diff
@@ -0,0 +1,134 @@
+"""Functions for parsing the inheritance for accumulating attributes.
+
+Accumulating attributes are both defined and can be added to and removed from further down the hierarchy.
+The hierarchy is also slightly different to the other fields as it is determined as 'breadth-first' in
+multiple inheritance, so targets at a lower level will always take precedence over targets at a higher level.
+"""
+from collections import deque
+from typing import Dict, List, Any, Deque
+
+ACCUMULATING_ATTRIBUTES = ["extra_labels", "macros", "device_has", "features", "components"]
+ALL_ACCUMULATING_ATTRIBUTES = ACCUMULATING_ATTRIBUTES + [
+    "extra_labels_remove",
+    "extra_labels_add",
+    "macros_remove",
+    "macros_add",
+    "device_has_remove",
+    "device_has_add",
+    "features_remove",
+    "features_add",
+    "components_remove",
+    "components_add",
+]
+
+
+def _targets_accumulate_hierarchy(all_targets_data: Dict[str, Any], target_name: str) -> List[dict]:
+    """List all ancestors of a target in order of accumulation inheritance (breadth-first).
+
+    Using a breadth-first traverse of the inheritance tree, return a list of targets in the
+    order of inheritance, starting with the target itself and finishing with its highest ancestor.
+
+    Args:
+        all_targets_data: a dictionary representation of all the data in a targets.json file
+        target_name: the name of the target we want to calculate the attributes for
+
+    Returns:
+        A list of dicts representing each target in the hierarchy.
+    """
+    accumulation_order: List[dict] = []
+
+    still_to_visit: Deque[dict] = deque()
+    still_to_visit.appendleft(all_targets_data[target_name])
+
+    while still_to_visit:
+        current_target = still_to_visit.popleft()
+        accumulation_order.append(current_target)
+        if "inherits" in current_target.keys():
+            for parent_target in current_target["inherits"]:
+                still_to_visit.append(all_targets_data[parent_target])
+
+    return accumulation_order
+
+
+def _add_attribute_element(
+    accumulator: Dict[str, Any], attribute_name: str, elements_to_add: List[Any]
+) -> Dict[str, Any]:
+    """Adds an attribute element to an attribute.
+
+    Args:
+        accumulator: a store of attributes to be updated
+        attribute_name: name of the attribute to update
+        elements_to_add: element to add to the attribute list
+
+    Returns:
+        The accumulator object with the new elements added
+    """
+    for element in elements_to_add:
+        accumulator[attribute_name].append(element)
+    return accumulator
+
+
+def _remove_attribute_element(
```

Comment:
This and `_add_attribute_element` look very similar at this stage but `_remove_attribute_element` will have to handle more complexity and so will change quite dramatically in the next PR.